ansible/library/bgp_route.py

```diff
@@ -0,0 +1,343 @@
+     *> 100.1.0.13/32    10.0.0.34                              0 65200 i
+     *> 100.1.0.15/32    10.0.0.34                              0 65200 i
+     *> 100.1.0.17/32    10.0.0.34                              0 64001 i
+     *> 192.168.10.97/32 10.0.0.34                              0 65200 62011 65507 ?
+     *> 192.168.10.112/32
+                         10.0.0.34                              0 65200 62011 65508 ?
+     *> 192.168.10.113/32
+                         10.0.0.34                              0 65200 62011 65508 ?
+     *> 192.168.10.128/32
+                         10.0.0.34                              0 65200 62011 65509 ?
+     *> 192.168.10.129/32
+                         10.0.0.34                              0 65200 62011 65509 ?
+     *> 192.168.10.144/32
+                         10.0.0.34                              0 65200 62011 65510 ?
+'''
+
+
+### TODO: No show ipv6 route entries parsing option, some show command will expose ipv6 routes, need continue working on ipv6 specific commands###
+
+
+class BgpRoutes(object):
+    '''
+        parsing bgp routing information
+    '''
+    def __init__(self):
+        self.facts = defaultdict(dict)
+        return
+
+    def get_facts(self):
+        return self.facts
+
+    def parse_bgp_route_adv(self, cmd, cmd_result):
+        '''
+        parse BGP routing facts of neighbor advertised routes
+        '''
+        regex_neighbor = re.compile('nei.*\s([0-9a-fA-F.:]+)\s.*')
+        #regix_iprange = re.compile('[0-9a-fA-F.:]+\/\d+')
+        if regex_neighbor.match(cmd):
+            neighbor = regex_neighbor.match(cmd).group(1)
+        else:
+            raise Exception('cannot find neighbor in "show ip  bgp ' + cmd + '"')
+        self.facts['bgp_route_neiadv']['neighbor'] = neighbor
+        ### so far parsing prifix, nexthop and paths only
+        header = 'Metric LocPrf Weight Path'
+        result_lines = cmd_result.split('\n')
+        table_start = False
+        re_path = re.compile('.*\s{2,}(\d+)\s((\d+\s)+)?([ie\?])$')
+        while len(result_lines) != 0:
+            line = result_lines.pop(0)
+            if not table_start:
+                if header in line:
+                    table_start = True
+                    continue
+            else:
+                ## only parse valid route entry, ignore if it's not marked as valid
+                if not re.match('^\*', line):
+                    continue
+                entry = dict()
+                fields = line.strip().split()
+                prefix = fields[1]
+                if len(fields) > 2:    ### route entry in one line
+                    nexthop = fields[2]
+                else:                  ### route entry in two lines
+                    line = result_lines.pop(0)
+                    nexthop = line.strip().split()[0]
+                paths = re_path.match(line).group(2)
+                origin = re_path.match(line).group(4)
```

Comment:
we only need one match

m = re_aspath.match(line)

aspaths = m.group(2)
origin = m.group(4)

what if it does not aspath, is it still group 4?

Reply:
weight = m.group(1)

---
In reply to: [145203052](https://github.com/Azure/sonic-mgmt/pull/284#discussion_r145203052) [](ancestors = 145203052)